### ISSUE TITLE
@threads with broadcast operator hangs or segfaults

### ISSUE THREAD

**Issue by tanmaykm** [JuliaLang/julia#14426]
The following operations results in a hang with all threads busy:

````
julia> using Base.Threads

julia> nthreads()
8

julia> a = [1:100;];

julia> @threads for i in 1:100
           c = a .+ a
       end
Killed
````

It causes segfault if gc is disabled.

````
julia> using Base.Threads

julia> gc_enable(false)
true

julia> a = [1:100;];

julia> @threads for i in 1:100
           c = a .+ a
       end

signal (11): Segmentation fault
while loading no file, in expression starting on line 0
while loading no file, in expresswhile loading no file, in expression starting on line 0
while loading no file, in expresswhile loading no file, in expresswhile loading no file, in expression starting on line 0
while loading no file, in expression starting on line 0
while loading no file, in expresswhile loading no file, in expresswhile loading no file, in expression starting on line 0
while loading no file, in expresswhile loading no file, in expression starting on line 0
while loading no file, in expression starting on line 0
while loading no file, in expresswhile loading no file, in expression starting on line 0
unknown function (ip: 0x7f45dd79bb3a)
unknown function (ip: 0x7f45dd79bb3a)
jl_expand at /data/Work/julia/sources/julia/usr/bin/../lib/libjulia.so (unknown line)
jl_expand at /data/Work/julia/sources/julia/usr/bin/../lib/libjulia.so (unknown line)
unknown function (ip: 0x7f45dd7c1cfe)
unknown function (ip: 0x7f45dd82d0c6)
unknown function (ip: 0x7f45dd8386e2)
unknown function (ip: 0x7f45dd7c1cfe)
unknown function (ip: 0x7f45dd83af48)
jl_toplevel_eval_in_warn at /data/Work/julia/sources/julia/usr/bin/../lib/libjulia.so (unknown line)
[inline] at broadcast.jl:3
unknown function (ip: 0x7f45dd82d0c6)
gen_broadcast_function at broadcast.jl:204
[inline] at dict.jl:711
broadcast! at broadcast.jl:243
unknown function (ip: 0x7f45dd82fd90)
unknown function (ip: 0x7f45dd8386e2)
jl_toplevel_eval_in_warn at /data/Work/julia/sources/julia/usr/bin/../lib/libjulia.so (unknown line)
...
````

Julia version:
````
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.0-dev+1824 (2015-12-16 08:25 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 79b2418* (0 days old master)
|__/                   |  x86_64-linux-gnu

julia> versioninfo()
Julia Version 0.5.0-dev+1824
Commit 79b2418* (2015-12-16 08:25 UTC)
Platform Info:
  System: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7-4700MQ CPU @ 2.40GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
````

**Comment by yuyichao** [comment]
Given it's in `jl_expand`, maybe try https://github.com/JuliaLang/julia/pull/14384 ?

**Comment by yuyichao** [comment]
I can reproduce this on current master but it seems to be fixed on #14190 with #14366, #14384 and #14391 applied.

Note that with #14384 this hits #14391 so I guess this can be closed when both are merged.

**Comment by yuyichao** [comment]
I just merged #14391 so I'll close this as a dup of https://github.com/JuliaLang/julia/issues/14354
